brain/morphology.cpp

```diff
@@ -0,0 +1,263 @@
+        }
+
+        brion::uint32_ts result;
+        for( uint32_t i = 0; i != _types->size(); ++i )
+        {
+            const brion::SectionType type = ( *_types )[i];
+            if( type > brion::SECTION_APICAL_DENDRITE + 1 )
+                LBWARN << "Unknown section type " << int(type) << std::endl;
+            else
+                if( bits[size_t( type )] )
+                    result.push_back( i );
+        }
+        return result;
+    }
+
+    brion::Vector4fs getSectionSamples( const size_t sectionID,
+                                        const brion::floats& points ) const
+    {
+        // If the section is the soma return directly the soma position.
+        if(( *_types )[sectionID] == brion::SECTION_SOMA )
+        {
+            brion::Vector4fs result;
+            for( size_t i = 0; i != points.size(); ++i )
+                result.push_back( _somaPosition );
+            return result;
+        }
+
+        const size_t start = ( *_sections )[sectionID][0];
+        const size_t end = sectionID == _sections->size() - 1 ?
+            _points->size() : ( *_sections )[sectionID + 1][0];
+
+        brion::Vector4fs result;
+
+        if( end <= start )
+        {
+            LBWARN << "Trying to sample broken morphology or empty section "
+                   << sectionID << std::endl;
+            return result;
+        }
+        // Dealing with the degenerate case of single point sections
+        if( start + 1 == end )
+            for( size_t i = 0; i != points.size(); ++i)
+                result.push_back(( *_points )[start] );
+
+        // Computing the accumulated length at each segment of the section
+        brion::floats accumLengths;
+        accumLengths.push_back(0);
+        for( size_t i = start; i != end - 1; ++i )
+        {
+            const brion::Vector4f& segmentStart = ( *_points )[i];
+            const brion::Vector4f& segmentEnd = ( *_points )[i + 1];
+            accumLengths.push_back( accumLengths.back() +
+                                    ( segmentEnd - segmentStart ).length( ));
+        }
+        const float totalLength = accumLengths.back();
+
+        for( brion::floats::const_iterator i = points.begin();
+             i != points.end(); ++i )
+        {
+            // Finding the segment the requested sampling position falls in
+            float length = std::max( 0.f, std::min( 1.f, *i )) * totalLength;
+            size_t index = 0;
+            while( accumLengths[index + 1] < length &&
+                   index < accumLengths.size() - 1 )
+                ++index;
+
+            // Interpolating the point
+            const float alpha = ( length - accumLengths[index] ) /
+                              ( accumLengths[index + 1] - accumLengths[index] );
+            const brion::Vector4f sample =
+                ( *_points )[start + index + 1] * alpha +
+                ( *_points )[start + index] * (1 - alpha );
+            result.push_back( sample );
```

Comment:
it's possible to reserve() the result? Or the amount of points is always small enough?